### PR TITLE
Mixer subsystem and commands

### PR DESCRIPTION
**Pull request by OhadLevine**
Get current is used for a specific command hilel asked for.
please cr

### CODE REVIEW

File: src/main/java/frc/robot/constants/RobotMap.java [hunk 1]
```diff
@@ -18,6 +18,7 @@
   // TODO: Set variables for hardware components
   
   public static class CAN {
+    public int MIXER_TALON;
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/constants/RobotMap.java:21]: remove empty line

File: src/main/java/frc/robot/subsystems/Mixer.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+public class Mixer extends SubsystemBase implements MoveableSubsystem {
+
+  private final WPI_TalonSRX motor;
+
+  public Mixer() {
+    motor = new WPI_TalonSRX(robotConstants.can.MIXER_TALON);
+  }
+
+  public void move(double power) {
+    motor.set(ControlMode.PercentOutput, power);
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/Mixer.java]: replace `motor.set(ControlMode.PercentOutput, power);` with `motor.set(power);`

File: src/main/java/frc/robot/subsystems/Mixer.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  }
+
+  public double getStall(){
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/Mixer.java]: add docs

**Review by EliyaFishman — CHANGES_REQUESTED** [review]
good job

File: src/main/java/frc/robot/subsystems/Mixer.java (hunk 2 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/subsystems/Mixer.java]: add `@Override`

**Review by EliyaFishman — CHANGES_REQUESTED** [review]

**Comment by OhadLevine** [comment]
done

**Review by EliyaFishman — CHANGES_REQUESTED** [review]
Move mixer to subsystems/mixer

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  @Override
+  public void initialize() {
+    startTime = Timer.getFPGATimestamp();
+  }
+
+  @Override
+  public void execute() {
+    if (Robot.mixer.getStall() < maxStall){
+      notOnTarget = Timer.getFPGATimestamp() - startTime;
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java]: why target? there isn't any vision in here. Also remove start time

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  /** gets a double for motor power */
+  public SpinMixer(double power, double maxStall) {
+    this(() -> power, maxStall);
+  }
+
+  @Override
+  public void initialize() {
+    startTime = Timer.getFPGATimestamp();
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java]: remove start time

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  @Override
+  public void execute() {
+    if (Robot.mixer.getStall() < maxStall){
+      notOnTarget = Timer.getFPGATimestamp() - startTime;
+    }
+    if (Timer.getFPGATimestamp() - startTime - notOnTarget > waitTime){
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java]: remove `- startTime`

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+      double speed = -power.getAsDouble();
+      Robot.mixer.move(speed);
+    }
+  }
+
+  @Override
+  public void end(boolean interrupted) {
+    Robot.mixer.move(0);
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java]: replace with `mixer.stopMove();`

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  @Override
+  public void execute() {
+    if (Robot.mixer.getStall() < maxStall){
+      notOnTarget = Timer.getFPGATimestamp() - startTime;
+    }
+    if (Timer.getFPGATimestamp() - startTime - notOnTarget > waitTime){
+      double speed = -power.getAsDouble();
+      Robot.mixer.move(speed);
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java]: replace two lines with `mixer.move(-power.getAsDouble());`

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+  /** gets a supplier for motor power */
+  public SpinMixer(DoubleSupplier power, double maxStall) {
+    addRequirements(Robot.mixer);
+    this.power = power;
+    this.maxStall = maxStall;
+  }
+
+  /** gets a double for motor power */
+  public SpinMixer(double power, double maxStall) {
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java]: we would probably want a constant maxStall. Move maxStall to RobotConstants

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java (hunk 4 above)
Comment (EliyaFishman) [reply src/main/java/frc/robot/subsystems/mixer/SpinMixer.java]: We want to just change direction when the stall is big enough. It should still speed even if that bellow

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
+package frc.robot.subsystems.mixer;
+
+import java.util.function.DoubleSupplier;
+
+import edu.wpi.first.wpilibj.Timer;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.Robot;
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java]: replace `import frc.robot.Robot;` with `import static frc.robot.Robot.mixer;`

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
+package frc.robot.subsystems.mixer;
+
+import java.util.function.DoubleSupplier;
+
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java]: remove empty line

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+import edu.wpi.first.wpilibj.Timer;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.Robot;
+
+/**
+ * Spins the mixer during the game for putting balls in the ejector, if the current is too big the motor power flips.
+ */
+public class SpinMixer extends CommandBase {
+  private DoubleSupplier power;
+  private double startTime;
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java]: remove start time as it's not needed

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  @Override
+  public void initialize() {
+    startTime = Timer.getFPGATimestamp();
+  }
+
+  @Override
+  public void execute() {
+    if (Robot.mixer.getStall() < maxStall){
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java]: replace `Robot.mixer.getStall()` with `mixer.getStall()`

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    }
+  }
+
+  @Override
+  public void end(boolean interrupted) {
+    Robot.mixer.move(0);
+  }
+
+  @Override
+  public boolean isFinished() {
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java]: delete method as you don't change anything from CommandBase

File: src/main/java/frc/robot/utils/DashboardDataContainer.java [hunk 15]
```diff
@@ -1,16 +1,21 @@
...
  */
 public class DashboardDataContainer {
     private DashboardController dashboardController;
-    
+
     public DashboardDataContainer() {
-        dashboardController = new DashboardController();
 
+        dashboardController = new DashboardController();
+        SmartDashboard.putNumber("Mixer Power", 0.0);
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/utils/DashboardDataContainer.java]: replace "Mixer Power" with "Mixer/Mixer Power".
This makes a subtable of mixer so shuffleboard is cleaner

File: src/main/java/frc/robot/utils/DashboardDataContainer.java [hunk 16]
```diff
@@ -1,16 +1,21 @@
...
+        SmartDashboard.putData("Spin Mixer", new SpinMixer(SmartDashboard.getNumber("Mixer Power", 0.0), 0));
```
Comment (EliyaFishman) [inline src/main/java/frc/robot/utils/DashboardDataContainer.java]: replace "Spin Mixer" "Mixer/Spin Mixer"

**Review by EliyaFishman — CHANGES_REQUESTED** [review]

**Review by EliyaFishman — APPROVED** [review]

File: src/main/java/frc/robot/constants/RobotConstants.java [hunk 17]
```diff
@@ -48,6 +49,9 @@
         public double TARGET_NOT_FOUND_WAIT_TIME;
     }
 
+    public static class MixerConstants {
+        public double MIXER_MAX_STALL;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/constants/RobotConstants.java]: add inverted

File: src/main/java/frc/robot/constants/RobotMap.java [hunk 18]
```diff
@@ -16,8 +16,8 @@
   public I2C i2c = new I2C(); 
 
   // TODO: Set variables for hardware components
-  
   public static class CAN {
+    public int MIXER_TALON;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/constants/RobotMap.java]: MIXER_TALON_SRX

File: src/main/java/frc/robot/subsystems/mixer/Mixer.java [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+/**
+ * This class holds all of the methods for the Mixer subsystem, which holds
+ * POWER CELLS in the robot for shooting.
+ */
+public class Mixer extends SubsystemBase implements MoveableSubsystem {
+
+  private final WPI_TalonSRX talon;
+
+  public Mixer() {
+    talon = new WPI_TalonSRX(robotConstants.can.MIXER_TALON);
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/subsystems/mixer/Mixer.java]: add coast, inverted

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+import static frc.robot.Robot.mixer;
+
+/**
+ * Spins the mixer during the game for putting balls in the ejector, if the
+ * current is too big the motor power flips.
+ */
+public class SpinMixer extends CommandBase {
+  private DoubleSupplier power;
+  private double lastTimeNotOnStall;
+  private double waitTime;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java]: set waitTime..?

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+  /** gets a double for motor power */
+  public SpinMixer(double power) {
+    this(() -> power);
+  }
+
+  @Override
+  public void execute() {
+    if (mixer.getStall() < robotConstants.mixerConstants.MIXER_MAX_STALL) {
+      lastTimeNotOnStall = Timer.getFPGATimestamp();
+      mixer.move(power.getAsDouble());
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java]: mixer.move(power.getAsDouble());
move to:

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+    }
+    if (Timer.getFPGATimestamp() - lastTimeNotOnStall > waitTime) {
+      mixer.move(-power.getAsDouble());
+    }
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java:54]: } else {
      mixer.move(power.getAsDouble()); //here
}

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+    this.power = power;
+  }
+
+  /** gets a double for motor power */
+  public SpinMixer(double power) {
+    this(() -> power);
+  }
+
+  @Override
+  public void execute() {
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java:42]: change method for flipping (wont work)

**Review by HilelAmiel — CHANGES_REQUESTED** [review]
few changes

File: src/main/java/frc/robot/constants/RobotConstants.java [hunk 24]
```diff
@@ -48,6 +49,10 @@
         public double TARGET_NOT_FOUND_WAIT_TIME;
     }
 
+    public static class MixerConstants {
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/constants/RobotConstants.java]: public static class MixerConstants {
move after drivetrain constans

File: src/main/java/frc/robot/subsystems/mixer/Mixer.java [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+
+import static frc.robot.Robot.robotConstants;
+
+/**
+ * This class holds all of the methods for the Mixer subsystem, which holds
+ * POWER CELLS in the robot for shooting.
+ */
+public class Mixer extends SubsystemBase implements MoveableSubsystem {
+
+  private final WPI_TalonSRX talon;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/subsystems/mixer/Mixer.java]: talon to talonSRX

File: src/main/java/frc/robot/subsystems/mixer/SpinMixer.java [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Spins the mixer during the game for putting balls in the ejector, if the
+ * current is too big the motor power flips.
+ */
+public class SpinMixer extends CommandBase {
+  private static final double WAIT_TIME = 0.2;
+  private DoubleSupplier power;
+  private double lastTimeNotOnStall;
+  private double changeDirectionTime;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/subsystems/mixer/SpinMixer.java]: changeDirectionTime what is for? not used

File: src/main/java/frc/robot/subsystems/mixer/Mixer.java [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+
+  public Mixer() {
+    talon = new WPI_TalonSRX(robotConstants.can.MIXER_TALON_SRX);
+    talon.setNeutralMode(NeutralMode.Coast);
+    talon.setInverted(robotConstants.mixerConstants.kIsInverted);
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/subsystems/mixer/Mixer.java]: we might want ramp limit...

**Review by HilelAmiel — CHANGES_REQUESTED** [review]
change last things

File: src/main/java/frc/robot/constants/RobotConstants.java (hunk 24 above)
Comment (HilelAmiel) [reply src/main/java/frc/robot/constants/RobotConstants.java]: you did that?

**Review by HilelAmiel — APPROVED** [review]

**Pull request merged**